File: src/typegraph/rust/typegraph/src/runtimes/deno.rs
// Copyright Metatype OÜ, licensed under the Elastic License 2.0.
// SPDX-License-Identifier: Elastic-2.0

use crate::{
    policy::Policy,
    t::{self, TypeBuilder, TypeDef},
    wasm::{
        self,
        core::RuntimeId,
        runtimes::{
            Effect, MaterializerDenoFunc, MaterializerDenoImport, MaterializerDenoPredefined,
        },
    },
    Result,
};

pub use super::ModuleImportOption as DenoImportOption;

#[derive(Debug)]
pub struct DenoRuntime {
    _id: RuntimeId,
}

impl DenoRuntime {
    pub fn new() -> Result<Self> {
        Ok(Self {
            _id: wasm::with_runtimes(|r, s| r.call_get_deno_runtime(s)),
        })
    }

    pub fn func<I, O>(&self, inp: I, out: O, options: DenoFuncOption) -> Result<TypeDef>
    where
        I: TypeBuilder,
        O: TypeBuilder,
    {
        let mat = MaterializerDenoFunc {
            code: options.code,
            secrets: options.secrets,
        };

        let mat = wasm::with_runtimes(|r, s| r.call_register_deno_func(s, &mat, options.effect))?;

        t::funcb(inp, out, mat)
    }

    pub fn import<I, O>(&self, inp: I, out: O, options: DenoImportOption) -> Result<TypeDef>
    where
        I: TypeBuilder,
        O: TypeBuilder,
    {
        let mat = MaterializerDenoImport {
            func_name: options.func_name,
            module: options.module,
            deps: options.deps,
            secrets: options.secrets,
        };

        let mat = wasm::with_runtimes(|r, s| r.call_import_deno_function(s, &mat, options.effect))?;

        t::funcb(inp, out, mat)
    }

    pub fn identity<I: TypeBuilder>(&self, inp: I) -> Result<TypeDef> {
        let mat = MaterializerDenoPredefined {
            name: "identity".to_string(),
        };

        let mat = wasm::with_runtimes(|r, s| r.call_get_predefined_deno_func(s, &mat))?;
        let inp = inp.build()?;

        t::funcb(inp.id(), inp.id(), mat)
    }

    pub fn policy(
        &self,
        name: &str,
        code: &str,
        secrets: impl IntoIterator<Item = impl ToString>,
    ) -> Result<Policy> {
        let mat = MaterializerDenoFunc {
            code: code.to_string(),
            secrets: secrets.into_iter().map(|s| s.to_string()).collect(),
        };

        let mat = wasm::with_runtimes(|r, s| r.call_register_deno_func(s, &mat, Effect::Read))?;

        Policy::create(name, mat)
    }

    pub fn import_policy(
        &self,
        name: Option<&str>,
        module: &str,
        func_name: &str,
        secrets: impl IntoIterator<Item = impl ToString>,
    ) -> Result<Policy> {
        let name = match name {
            Some(value) => value.to_string(),
            None => format!("__imp_{module}_{func_name}")
                .chars()
                .map(|ch| if ch.is_alphanumeric() { ch } else { '_' })
                .collect(),
        };

        let mat = MaterializerDenoImport {
            func_name: func_name.to_string(),
            module: module.to_string(),
            deps: Vec::default(),
            secrets: secrets.into_iter().map(|s| s.to_string()).collect(),
        };

        let mat = wasm::with_runtimes(|r, s| r.call_import_deno_function(s, &mat, Effect::Read))?;

        Policy::create(&name, mat)
    }

    // TODO
    // pub fn static(&self, ...) -> Result<TypeId> {}
}

#[derive(Debug, Default)]
pub struct DenoFuncOption {
    pub code: String,
    pub secrets: Vec<String>,
    pub effect: Effect,
}

impl DenoFuncOption {
    pub fn new(code: &str) -> Self {
        Self {
            code: code.to_string(),
            ..Default::default()
        }
    }

    pub fn secrets(mut self, secrets: impl IntoIterator<Item = impl ToString>) -> Self {
        self.secrets = secrets.into_iter().map(|s| s.to_string()).collect();
        self
    }

    pub fn effect(mut self, effect: Effect) -> Self {
        self.effect = effect;
        self
    }
}

File: src/typegraph/python/typegraph/runtimes/deno.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

import json
import re
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, List, Optional, Union, overload

from typegraph.gen.runtimes import (
    Effect,
    MaterializerDenoFunc,
    MaterializerDenoImport,
    MaterializerDenoPredefined,
    MaterializerDenoStatic,
)
from typegraph.policy import Policy
from typegraph.runtimes.base import Materializer, Runtime
from typegraph.sdk import runtimes
from typegraph.utils import ResolvedModule, resolve_module_params
from typegraph.utils import Module

# from typegraph.sdk import sdk_utils


if TYPE_CHECKING:
    from typegraph import t


DenoModule = Module


class DenoRuntime(Runtime):
    def __init__(self):
        super().__init__(runtimes.get_deno_runtime())

    def static(self, out: "t.typedef", value: Any):
        from typegraph import t

        mat_id = runtimes.register_deno_static(
            MaterializerDenoStatic(json.dumps(value)), out._id
        )

        return t.func(
            t.struct(),
            out,
            StaticMat(
                mat_id,
                value=value,
                effect="read",
            ),
        )

    def func(
        self,
        inp: "t.struct",
        out: "t.typedef",
        *,
        code: str,
        secrets: Optional[List[str]] = None,
        effect: Optional[Effect] = None,
    ):
        secrets = secrets or []
        effect = effect or "read"
        mat_id = runtimes.register_deno_func(
            MaterializerDenoFunc(code=code, secrets=secrets),
            effect,
        )

        from typegraph import t

        return t.func(
            inp, out, FunMat(mat_id, code=code, secrets=secrets, effect=effect)
        )

    @overload
    def import_(
        self,
        inp: "t.struct",
        out: "t.typedef",
        *,
        module: str,
        name: str,
        deps: List[str],
        effect: Optional[Effect],
        secrets: Optional[List[str]],
    ): ...

    @overload
    def import_(
        self,
        inp: "t.struct",
        out: "t.typedef",
        *,
        module: ResolvedModule,
    ): ...

    def import_(
        self,
        inp: "t.struct",
        out: "t.typedef",
        *,
        module: Union[str, ResolvedModule],
        name: Optional[str] = None,
        deps: Optional[List[str]] = None,
        effect: Optional[Effect] = None,
        secrets: Optional[List[str]] = None,
    ):
        effect = effect or "read"
        secrets = secrets or []
        resolved = resolve_module_params(module, name, deps)
        mat_id = runtimes.import_deno_function(
            MaterializerDenoImport(
                func_name=resolved.func_name,
                module=resolved.module,
                secrets=secrets,
                deps=resolved.deps,
            ),
            effect,
        )

        from typegraph import t

        return t.func(
            inp,
            out,
            ImportMat(
                id=mat_id,
                name=resolved.func_name,
                module=resolved.module,
                secrets=secrets,
                effect=effect,
            ),
        )

    def identity(self, inp: "t.struct") -> "t.func":
        from typegraph import t

        res = runtimes.get_predefined_deno_func(
            MaterializerDenoPredefined(name="identity", param=None)
        )

        return t.func(
            inp,
            inp,
            PredefinedFunMat(id=res, name="identity", effect="read"),
        )

    def fetch_context(self, output_shape: Union["t.struct", None] = None):
        """
        Utility for fetching the current request context
        """
        return_value = (
            "context" if output_shape is not None else "JSON.stringify(context)"
        )
        from typegraph import t

        return self.func(
            inp=t.struct(),
            out=output_shape or t.json(),
            code=f"(_, {{ context }}) => {return_value}",
        )

    def policy(
        self, name: str, code: str, secrets: Optional[List[str]] = None
    ) -> Policy:
        secrets = secrets or []
        mat_id = runtimes.register_deno_func(
            MaterializerDenoFunc(code=code, secrets=secrets),
            "read",
        )

        return Policy.create(
            name,
            mat_id,
        )

    @overload
    def import_policy(
        self,
        *,
        name: str,
        module: ResolvedModule,
        secrets: Optional[List[str]],
    ) -> Policy: ...

    @overload
    def import_policy(
        self,
        *,
        name: Optional[str],
        func_name: str,
        module: str,
        deps: List[str],
        secrets: Optional[List[str]],
    ) -> Policy: ...

    def import_policy(
        self,
        *,
        name: Optional[str] = None,
        module: Union[str, ResolvedModule],
        func_name: Optional[str] = None,
        deps: Optional[List[str]] = None,
        secrets: Optional[List[str]] = None,
    ) -> Policy:
        resolved = resolve_module_params(module, func_name, deps)
        name = name or re.sub(
            "[^a-zA-Z0-9_]", "_", f"__imp_{resolved.module}_{resolved.func_name}"
        )

        res = runtimes.import_deno_function(
            MaterializerDenoImport(
                func_name=resolved.func_name,
                module=resolved.module,
                deps=resolved.deps,
                secrets=secrets or [],
            ),
            "read",
        )

        return Policy.create(name, res)


@dataclass
class FunMat(Materializer):
    code: str
    secrets: List[str]


@dataclass
class StaticMat(Materializer):
    value: Any


@dataclass
class ImportMat(Materializer):
    module: str
    name: str
    secrets: List[str]


@dataclass
class PredefinedFunMat(Materializer):
    name: str
